### ISSUE TITLE
Validating Circular Associations, replaces validations with generic ActiveRecord::RecordNotSaved error

### ISSUE THREAD

**Issue by TaylorHG** [rails/rails#21319]
Given:

    def Node
      belongs_to :parent, class_name: :Node
      has_many :children, class_name: :Node, foreign_key: :parent_id
    end

I am trying to create a Validation to make sure that a Node cannot be it's own parent, or it's parent's parent, etc.

I got this:

    # Custom validator to check if the Parent Node is related to the current node. Avoids that ugly self-association loop.
    #
    class NodeParentValidator < ActiveModel::Validator
      def validate(node)
        @node = node
    
        unless validate_recursive(node)
          node.errors[:parent_id] << "Node can't be it's own parent"
        end
    
      end
    
      def validate_recursive(node)
        # If the validating node doesn't have a parent, we return true because we're either facing a node that doesn't have any parent
        # or we got to the point that we need to stop the recursive loop.
        return true if node.parent.nil?
        # If the validating node is the same as the current "parent", we found the self-association loop. This is bad.
        return false if @node.id == node.parent_id
        # While we don't get to the end of the association, keep jumping.
        return false unless validate_recursive(node.parent)
        return true
      end
    
    end

The Validation catches a record being assigned to itself, and when Rails called the assign_attributes method, I get a 422, but it doesn't have my Validation in it! Instead I get an ugly HTML validation error like this:

    ActiveRecord::RecordNotSaved (Failed to replace children because one or more of the new records could not be saved.)

So if Rails can't save it's associated records, Rails returns it's own Error (the one in a code block above), but if my Record is associated to itself, I get a large problem. Even if I stop the Node from validating it's associated children/parents, I still get the RecordNotSaved error, rather than the error from my own Validation I created.

As long as the record I am trying to save ITSELF has an error, then Rails replaces my 422 with the error above. And that is just plain bad. I want a JSON response error so my client knows what exactly went wrong.

TLDR; If a record is associated with itself, and it has a validation error, then Validation errors created in that record's own Validations will never be displayed, replaced instead by Rails RecordNotSaved error.

I find it hard to believe that no one else has run into this problem, am I missing something or is this truly a bug?

**Comment by TaylorHG** [comment]
Anyone?

**Comment by rafaelfranca** [comment]
Could you create a reproduction script with this template https://github.com/rails/rails/blob/master/guides/bug_report_templates/active_record_master.rb?

**Comment by TaylorHG** [comment]
Thanks for responding! Here is the test: https://gist.github.com/TaylorHG/d0736567612beee22c97

**Comment by zerothabhishek** [comment]
Hello @TaylorHG 


I tried the test code you provided, and the statement that fails is this one:

    comment.update(parent_id: comment.id, child_ids: ["#{comment.id}"])

And as far as I figured out, the following would also fail with same error:

     node.child_ids = [node.id]
     node.children = [node]

In fact the above would also raise the error for any validation failing for the child nodes.

The reason is [this part of activerecord code](https://github.com/rails/rails/blob/master/activerecord/lib/active_record/associations/collection_association.rb#L553), which is called when we try replacements on associated collections. I could not find any documentation or rationale for throwing the exception as we see (instead of perhaps reverting the transaction quietly).

The other scenarios work as expected

    node.children << node       # => false
    node.errors.full_messages   #=> Parent Node can't be it's own parent
    node.errors.clear
    node.update(parent_id: node.id)  #=> false
    node.errors.full_messages        #=> Parent Node can't be it's own parent

Hope this is useful.

**Comment by TaylorHG** [comment]
Hey @zerothabhishek, sorry it took so long to respond.

Thanks for checking out the issue. I should have noted in my test that this issue also occurs with a circular *chain* rather than just a single record relating to itself, [here](https://gist.github.com/TaylorHG/978168cbd32b2af85de0) is a gist to show that behavior.

As for the issue itself, do you think it is something that warrants fixing, or at least documentation? I am rather surprised that nobody else appears to be working with validation errors in circular relationships.
